mautic/mautic
# Optimizing contact batch api

~~⚠️ This PR needs commits from https://github.com/mautic/mautic/pull/11272 to work so that one needs to be merged first.~~

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y
| Automated tests included? | Y
| Related user documentation PR URL | /
| Related developer documentation PR URL | /
| Issues addressed (#s or URLs) | /
| BC breaks? | N
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

This PR contains several performance optimizations to the `POST api/contacts/batch/new` API endpoint.

Adding indexes that help a little.

This PR also optimizes how the form fetches campaigns. From this query:
```sql
SELECT cl.company_id, cl.date_added as date_associated, cl.is_primary, comp.*
FROM mautic_companies_leads cl
INNER JOIN mautic_companies comp ON comp.id = cl.company_id
WHERE cl.lead_id = ?
```
to this one:
```sql
SELECT cl.company_id FROM mautic_companies_leads cl WHERE cl.lead_id = ?
```

This PR is also caching custom field results wherever possible. It used to be caching the Paginator object instead of the actual result so it was always fetching new rows instead of the results. This reduced contact field queries from 381 to 9 queries per 100 contacts in the API request payload.

*All the changes we've done so far on this ticket speeded up the API request containing 100 contacts from 8.5 seconds to 7.4 seconds (13%).*

We were hoping for more but there is no clear bottleneck in this API request. Just a lot of small improvements can be made. One can save a few hundred milliseconds so it's lots of changes with small measurable improvement.

#### Steps to test this PR:
1. Make sure the migrations execute without any issue. Improvement to the API endpoint should be visible.
2. Make sure you can still add/edit campaigns in the contact edit form (it was optimized)
3. Make sure you can still edit a contact via UI. Focus on companies, stages and owner select fields as those were touched in this PR.

#### Other areas of Mautic that may be affected by the change:
1. The index is added to the frequency rules table and lead fields table but it should not have any measurable effect on storing new records to those tables.

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/11277"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
